tasks/hypev.py

```diff
@@ -3,95 +3,249 @@
 
         if self.vocab is None:
-            vocab = Vocabulary(s0 + s1)
+            vocab = Vocabulary(s0 + s1)  # FIXME: lower?
         else:
             vocab = self.vocab
 
         si0 = vocab.vectorize(s0, spad=self.s0pad)
         si1 = vocab.vectorize(s1, spad=self.s1pad)
         f0, f1 = nlp.sentence_flags(s0, s1, self.s0pad, self.s1pad)
         gr = graph_input_anssel(si0, si1, y, f0, f1, s0, s1)
+        gr, y = self.merge_questions(gr)
+        if save_cache:
+            with open(cache_filename, "wb") as f:
+                pickle.dump((s0, s1, y, vocab, gr), f)
+                print("save")
 
         return (gr, y, vocab)
 
     def build_model(self, module_prep_model, do_compile=True):
-        if self.c['ptscorer'] is None:
-            # non-neural model
-            return module_prep_model(self.vocab, self.c)
-
-        # ranking losses require wide output domain
-        oact = 'sigmoid' if self.c['loss'] == 'binary_crossentropy' else 'linear'
 
-        model = self.prep_model(module_prep_model, oact=oact)
+        model = build_model(self.emb, self.vocab, module_prep_model, self.c)
 
         for lname in self.c['fix_layers']:
             model.nodes[lname].trainable = False
 
         if do_compile:
-            model.compile(loss={'score': self.c['loss']}, optimizer=self.c['optimizer'])
+            model.compile(loss={'score': self.c['loss']}, optimizer=self.c['opt'])
         return model
 
     def fit_callbacks(self, weightsf):
-        return [HypEvCB(self.grv['si0'], self.grv),
-                ModelCheckpoint(weightsf, save_best_only=True, monitor='acc', mode='max'),
-                EarlyStopping(monitor='acc', mode='max', patience=4)]
+        return [ModelCheckpoint(weightsf, save_best_only=True, monitor='val_loss', mode='min'),
+                EarlyStopping(monitor='val_loss', mode='min', patience=10)]
 
     def eval(self, model):
         res = []
         for gr, fname in [(self.gr, self.trainf), (self.grv, self.valf), (self.grt, self.testf)]:
             if gr is None:
                 res.append(None)
                 continue
-            ypred = model.predict(gr)['score'][:,0]
-            res.append(ev.eval_hypev(ypred, gr['si0'], gr['score'], fname))
+            loss, acc = model.evaluate(gr, show_accuracy=True)
+            res.append(YesNoRes(loss, acc))
         return tuple(res)
 
     def res_columns(self, mres, pfx=' '):
         """ Produce README-format markdown table row piece summarizing
         important statistics """
-        return('%s%.6f |%s%.6f |%s%.6f'
-               % (pfx, mres[self.trainf]['QAccuracy'],
-                  pfx, mres[self.valf]['QAccuracy'],
-                  pfx, mres[self.testf].get('QAccuracy', np.nan)))
+        return('%s%.6f |%s%.6f |%s%.6f '
+               % (pfx, mres[self.trainf]['Precision'],
+                  pfx, mres[self.valf]['Precision'],
+                  pfx, mres[self.testf].get('Precision', np.nan)))
```

Comment:
So, is it precision or accuracy? I thought .evaluate() reports accuracy, and that's probably what we want too.